src/Build/BackEnd/Components/RequestBuilder/TargetUpToDateChecker.cs

```diff
@@ -1075,6 +1085,175 @@ internal static bool IsAnyOutOfDate<T>(out DependencyAnalysisLogDetail dependenc
             return false;
         }
 
+#if FEATURE_ENUMERATION
+        internal static bool IsAnyOutOfDateWip<T>(out DependencyAnalysisLogDetail dependencyAnalysisDetailEntry, string projectDirectory, IList<T> inputs, IList<T> outputs)
+        {
+            ErrorUtilities.VerifyThrow((inputs.Count > 0) && (outputs.Count > 0), "Need to specify inputs and outputs.");
+            ValidateTargetItemsType(inputs);
+            ValidateTargetItemsType(outputs);
+
+            // Algorithm: walk through all the outputs to find the oldest output
+            //            walk through the inputs as far as we need to until we find one that's newer (if any)
+
+            // PERF -- we could change this to ensure that we walk the shortest list first (because we walk that one entirely): 
+            //         possibly the outputs list isn't actually the shortest list. However it always is the shortest
+            //         in the cases I've seen, and adding this optimization would make the code hard to read.
+            (string filePath, DateTime fileTime) oldestOutput = GetOldestFile(projectDirectory, outputs);
+            (string filePath, DateTime fileTime) newerInput = GetNewerFile(projectDirectory, inputs, oldestOutput.fileTime);
```

Comment:
If there's potential overlap between input and output files (when they share the same directory), then it would be worth it to reuse data between the two methods